### PR TITLE
[Merged by Bors] - feat: port Topology.MetricSpace.Polish

### PR DESCRIPTION
**Pull request by urkud**
---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/Topology/MetricSpace/HausdorffDistance.lean [hunk 1]
```diff
@@ -591,6 +591,12 @@ theorem _root_.IsClosed.not_mem_iff_infDist_pos (h : IsClosed s) (hs : s.Nonempt
   simp [h.mem_iff_infDist_zero hs, infDist_nonneg.gt_iff_ne]
 #align is_closed.not_mem_iff_inf_dist_pos IsClosed.not_mem_iff_infDist_pos
 
+theorem continuousAt_inv_infDist_pt (h : x ∉ closure s) : ContinuousAt (fun x ↦ (infDist x s)⁻¹) x := by
```
Comment (urkud) [inline Mathlib/Topology/MetricSpace/HausdorffDistance.lean:594]:
```suggestion
theorem continuousAt_inv_infDist_pt (h : x ∉ closure s) :
    ContinuousAt (fun x ↦ (infDist x s)⁻¹) x := by
```

File: Mathlib/Topology/Constructions.lean [hunk 2]
```diff
@@ -1395,19 +1395,20 @@ theorem pi_generateFrom_eq_finite {π : ι → Type _} {g : ∀ a, Set (Set (π
     by_cases a ∈ i <;> simp [*]
 #align pi_generate_from_eq_finite pi_generateFrom_eq_finite
 
+theorem induced_to_pi {X : Type _} (f : X → ∀ i, π i) :
```
Comment (urkud) [inline Mathlib/Topology/Constructions.lean:1398]:
```suggestion
-- porting note: new lemma
theorem induced_to_pi {X : Type _} (f : X → ∀ i, π i) :
```

File: Mathlib/Topology/MetricSpace/HausdorffDistance.lean [hunk 3]
```diff
@@ -591,6 +591,13 @@ theorem _root_.IsClosed.not_mem_iff_infDist_pos (h : IsClosed s) (hs : s.Nonempt
   simp [h.mem_iff_infDist_zero hs, infDist_nonneg.gt_iff_ne]
 #align is_closed.not_mem_iff_inf_dist_pos IsClosed.not_mem_iff_infDist_pos
 
+theorem continuousAt_inv_infDist_pt (h : x ∉ closure s) :
```
Comment (urkud) [inline Mathlib/Topology/MetricSpace/HausdorffDistance.lean:594]:
```suggestion
-- porting note: new lemma
theorem continuousAt_inv_infDist_pt (h : x ∉ closure s) :
```

**Comment by urkud** [comment]
@eric-wieser Why did you force-push?

**Comment by eric-wieser** [comment]
Please run `git rebase origin/master` to handle the force push

**Comment by urkud** [comment]
@eric-wieser Fixed, thanks!

File: Mathlib/Topology/MetricSpace/HausdorffDistance.lean [hunk 4]
```diff
@@ -591,6 +591,14 @@ theorem _root_.IsClosed.not_mem_iff_infDist_pos (h : IsClosed s) (hs : s.Nonempt
   simp [h.mem_iff_infDist_zero hs, infDist_nonneg.gt_iff_ne]
 #align is_closed.not_mem_iff_inf_dist_pos IsClosed.not_mem_iff_infDist_pos
 
+-- porting note: new lemma
+theorem continuousAt_inv_infDist_pt (h : x ∉ closure s) :
```
Comment (eric-wieser) [inline Mathlib/Topology/MetricSpace/HausdorffDistance.lean:595]: Is thisre precedence for this `pt` name?

File: Mathlib/Topology/MetricSpace/Polish.lean [hunk 5]
```diff
@@ -197,287 +195,204 @@ theorem exists_polishSpace_forall_le {ι : Type _} [Countable ι] [t : Topologic
...
     exact this
   have L : @ClosedEmbedding _ _ (T.induced f) T f := by
-    constructor
+    refine @ClosedEmbedding.mk _ _ (T.induced f) T f ?_ ?_
     · exact K
     · exact f_closed
   exact @ClosedEmbedding.polishSpace _ _ (T.induced f) T (by infer_instance) _ L
 #align polish_space.exists_polish_space_forall_le PolishSpace.exists_polishSpace_forall_le
 
+end PolishSpace
```
Comment (eric-wieser) [inline Mathlib/Topology/MetricSpace/Polish.lean:236]: Where did this come from? Adding new sections during porting makes it harder to work out how to forward-port things.

File: Mathlib/Topology/MetricSpace/Polish.lean [hunk 6]
```diff
@@ -197,287 +195,204 @@ theorem exists_polishSpace_forall_le {ι : Type _} [Countable ι] [t : Topologic
...
+-- porting note: was @[nolint has_nonempty_instance]
+def CompleteCopy {α : Type _} [MetricSpace α] (s : Opens α) : Type _ := s
+#align polish_space.complete_copy TopologicalSpace.Opens.CompleteCopyₓ
+
+namespace CompleteCopy
+
+/-- A distance on an open subset `s` of a metric space, designed to make it complete.  It is given
+by `dist' x y = dist x y + |1 / dist x sᶜ - 1 / dist y sᶜ|`, where the second term blows up close to
+the boundary to ensure that Cauchy sequences for `dist'` remain well inside `s`. -/
+instance instDist : Dist (CompleteCopy s) where
```
Comment (eric-wieser) [inline Mathlib/Topology/MetricSpace/Polish.lean:267]: Shouldn't this have a porting note? It wasn't an instance in Lean 3.

File: Mathlib/Topology/MetricSpace/HausdorffDistance.lean (hunk 4 above)
Comment (urkud) [reply Mathlib/Topology/MetricSpace/HausdorffDistance.lean:595]: Yes, we have `continuous_infDist_pt`. Probably, we also have (at least, in Lean 3) a lemma saying that it's continuous both in pt and set.

File: Mathlib/Topology/MetricSpace/Polish.lean (hunk 5 above)
Comment (urkud) [reply Mathlib/Topology/MetricSpace/Polish.lean:236]: See porting note in the next docstring.

File: Mathlib/Topology/MetricSpace/Polish.lean [hunk 7]
```diff
@@ -0,0 +1,428 @@
...
```
Comment (semorrison) [inline Mathlib/Topology/MetricSpace/Polish.lean:267]:
```suggestion
-- Porting note: in mathlib3 this was only a local instance.
instance instDist : Dist (CompleteCopy s) where
```

File: Mathlib/Topology/MetricSpace/Polish.lean [hunk 8]
```diff
@@ -0,0 +1,428 @@
...
+instance : T0Space (CompleteCopy s) := inferInstanceAs (T0Space s)
+
+/-- A metric space structure on a subset `s` of a metric space, designed to make it complete
+if `s` is open. It is given by `dist' x y = dist x y + |1 / dist x sᶜ - 1 / dist y sᶜ|`, where the
+second term blows up close to the boundary to ensure that Cauchy sequences for `dist'` remain well
+inside `s`.
+
+Porting note: the definition changed to ensure that the `TopologicalSpace` structure on
+`TopologicalSpace.Opens.CompleteCopy s` is definitionally equal to the original one. -/
+instance instMetricSpace : MetricSpace (CompleteCopy s) := by
```
Comment (semorrison) [inline Mathlib/Topology/MetricSpace/Polish.lean:290]:
```suggestion
-- Porting note: in mathlib3 this was only a local instance.
instance instMetricSpace : MetricSpace (CompleteCopy s) := by
```

**Comment by semorrison** [comment]
bors merge

**Pull request closed without merging**